### PR TITLE
XBMCTinyXML2: Strip trailing '\0' when writing to file

### LINKED ISSUE #24135: xml issue 21.0-beta1 (20.90.801) installed via flatpak
`<favourites>
    <favourite name="Albert Park Circuit (Melbourne" thumb="http://127.0.0.1:57342/picture/2c7651463a04438aa12693e7ed9c7ec1/p-13751-0-p-4ecd55858cb06e2a8b73427c620db8b6">PlayMedia("http://127.0.0.1:57342/movies/2c7651463a04438aa12693e7ed9c7ec1/13743/m-13751-54a2a7b26582f68357d2add1283bdb5c-21-180-0-1-0-0-0-0-h264-8809759-ac3-384000--Albert_Park_Circuit__Melbourne__2012.mkv")</favourite>
</favourites>
\00`

tail \00
-> favourites.xml 

probably same for all xmls (sources.xml etc)

### PR DESCRIPTION
**Pull request by neo1973**
## Description
When writing the serialized XML to a file we don't want to add the terminating `'\0'` of the string to file.

Fixes #24135.

## Motivation and context
Text editors, validators, etc. don't expect the terminating null. Example:
```
$ xmllint .kodi/userdata/favourites.xml 
.kodi/userdata/favourites.xml:4: parser error : Extra content at the end of the document

^
```

## How has this been tested?
Add something to the favourites in Kodi. Without this change `favourites.xml` has an additional null byte at the end, with this change it's gone.

## What is the effect on users?
Valid XML files.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

**Comment by enen92** [comment]
> Can you say why this was not a problem before omega? What change in kodi code base led to the problem?

TinyXML2 was only introduced for omega as the sucessor of the unmaintained TinyXML. We bundle the two until we manage to get rid of all usages of the former in the codebase.

**Comment by neo1973** [comment]
We switched from TinyXML to TinyXML-2 a couple of month ago. The switch also isn't complete yet so some files are still processed with the old TinyXML and they don't have this issue.

File: xbmc/utils/XBMCTinyXML2.cpp [hunk 1]
```diff
@@ -57,8 +57,8 @@ bool CXBMCTinyXML2::SaveFile(const std::string& filename) const
   {
     tinyxml2::XMLPrinter printer;
     Accept(&printer);
-    bool suc =
-        file.Write(printer.CStr(), printer.CStrSize()) == static_cast<ssize_t>(printer.CStrSize());
+    const ssize_t sizeToWrite = printer.CStrSize() - 1; // strip trailing '\0'
```
Comment (fuzzard) [inline xbmc/utils/XBMCTinyXML2.cpp:60]: I'm not sure where the trailing null is coming from, but this seems really generic and might bite us in the ass if the actual cause of the null is resolved later. Better to check it's actually a null as part the process?

Comment (neo1973) [reply xbmc/utils/XBMCTinyXML2.cpp:60]: The trailing null is in the documentation so I'd say we can safely remove it without further checking:
```cpp
    /**
    	If in print to memory mode, return the size
    	of the XML file in memory. (Note the size returned
    	includes the terminating null.)
    */
    int CStrSize() const {
        return _buffer.Size();
    }
```

Comment (fuzzard) [reply xbmc/utils/XBMCTinyXML2.cpp:60]: All good then

**Pull request merged**